nodejs/aws-infra/experimental/clusterAutoScaling.ts

```diff
@@ -0,0 +1,566 @@
+import * as pulumi from "@pulumi/pulumi";
+
+import * as mod from ".";
+
+import { Overwrite, sha1hash } from "./../utils";
+
+export class ClusterAutoScalingLaunchConfiguration extends pulumi.ComponentResource {
+    public readonly instance: aws.ec2.LaunchConfiguration;
+
+    public readonly cluster: mod.Cluster;
+    /**
+     * Optional file system to mount.  Use [cluster.createFileSystem] to create an instance of this.
+     */
+    public readonly fileSystem?: mod.ClusterFileSystem;
+
+    public readonly instanceProfile: aws.iam.InstanceProfile;
+
+    /**
+     * Name to give the auto-scaling-group's cloudformation stack name.
+     */
+    public readonly stackName: pulumi.Output<string>;
+
+    constructor(name: string,
+                args: ClusterAutoScalingLaunchConfigurationArgs,
+                opts: pulumi.ComponentResourceOptions = {}) {
+        super("aws-infra.x.ClusterAutoScalingLaunchConfiguration", name, args, opts);
+
+        const parentOpts = { parent: this };
+
+        const cluster = args.cluster;
+
+        // Create the full name of our CloudFormation stack here explicitly. Since the CFN stack
+        // references the launch configuration and vice-versa, we use this to break the cycle.
+        // TODO[pulumi/pulumi#381]: Creating an S3 bucket is an inelegant way to get a durable,
+        // unique name.
+        const stackName = pulumi.output(args.stackName!) || new aws.s3.Bucket(name, {}, parentOpts).id;
+
+        // Use the instance provided, or create a new one.
+        const instanceProfile = getInstanceProfile(name, args, parentOpts);
+        const fileSystem = args.fileSystem;
+
+        const instance = new aws.ec2.LaunchConfiguration(
+            name, {
+            ...args,
+            imageId: getEcsAmiId(args.ecsOptimizedAMIName),
+            instanceType: pulumi.output(args.instanceType).apply(t => t || "t2.micro"),
+            keyName: args.keyName,
+            iamInstanceProfile: instanceProfile.id,
+            enableMonitoring: pulumi.output(args.enableMonitoring).apply(b => b !== undefined ? b : true),
+            placementTenancy: pulumi.output(args.placementTenancy).apply(t => t || "default"),
+            rootBlockDevice: pulumi.output(args.rootBlockDevice).apply(d => d || defaultRootBlockDevice),
+            ebsBlockDevices: pulumi.output(args.ebsBlockDevices).apply(d => d || defaultEbsBlockDevices),
+            securityGroups: pulumi.output(args.securityGroups).apply(g => g || [ cluster.instanceSecurityGroup.id ]),
+            userData: getInstanceUserData(cluster, args, stackName),
+        }, parentOpts);
+
+        this.instance = instance;
+        this.cluster = cluster;
+        this.stackName = stackName;
+        this.instanceProfile = instanceProfile;
+        this.fileSystem = fileSystem;
+
+        this.registerOutputs({
+            instance,
+            cluster,
+            stackName,
+            instanceProfile,
+            fileSystem,
+        });
```

Comment:
general pattern i'm following is to do all initialization of resources first, then assign the resources to 'this', then 'registerOutputs'.